clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -581,12 +582,24 @@ class DiagDeviceFunction : public RecursiveASTVisitor<DiagDeviceFunction> {
             << Sema::KernelCallRecursiveFunction;
       }
 
-      if (const CXXMethodDecl *Method = dyn_cast<CXXMethodDecl>(Callee))
+      if (const CXXMethodDecl *Method = dyn_cast<CXXMethodDecl>(Callee)) {
         if (Method->isVirtual() &&
             !SemaRef.getLangOpts().SYCLAllowVirtualFunctions)
           SemaRef.Diag(e->getExprLoc(), diag::err_sycl_restrict)
               << Sema::KernelCallVirtualFunction;
 
+        QualType Ty =
+            SemaRef.getASTContext().getRecordType(Method->getParent());
+        if (isSyclType(Ty, SYCLTypeAttr::esimd_simd_or_simd_mask) &&
+            Method->getNameAsString() == "data_ref") {
+          for (auto Use : SemaRef.getASTContext().getParents(*e)) {
```

Comment:
Glad I had you review this to catch that, thanks!

Yes, i think we need to do this in FE. Once we get to the middle end, the function call is always inlined and we can't reliably detect the bad case anymore.. Can you point me to an example or test that uses deferred diagnostics, I'm not familiar with then.

Reply:
For deferred diagnostics in the source code, start with `SYCLDiagIfDeviceCode(...)` in `SemaSYCL.cpp` and you can trace what we do there.

For a test that uses it, see e.g. `SemaSYCL/sycl-restrict.cpp`.

Reply:
thanks, i think i see what to do. ill update this pr once i have something

Reply:
nevermind, im stuck again. i can't find anywhere where we have the expr of the function call when doing deferred diagnostics, which is what need here since we only want to allow this function call for a particular use (directly inside an asm stmt). did i miss something?

Reply:
@premanandrao @elizabethandrews any ideas on the above? thanks!